framework/Models/HybridModel.py

```diff
@@ -0,0 +1,621 @@
+      outputMetrics = self.cvInstance.evaluateSample([romInfo['Instance'], self.tempTargetEvaluation], samplerType, kwargs)[1]
+      converged = self.isRomConverged(outputMetrics)
+      romInfo['Converged'] = converged
+      if converged:
+        romInfo['Instance'].train(self.tempTargetEvaluation)
+        self.raiseADebug("ROM ", romInfo['Instance'].name, " is converged!")
+
+  def isRomConverged(self, outputDict):
+    """
+      This function will check the convergence of rom
+      @ In, outputDict, dict, dictionary contains the metric information
+      @ Out, None
+    """
+    converged = True
+    for targetName, metricInfo in outputDict.items():
+      for metricName, metricValues in metricInfo.items():
+        if max(metricValues) > self.romConvergence:
+          self.raiseADebug("maximum error: ", max(metricValues), " is greater than the given tolerance")
+          converged = False
+    return converged
+
+  def checkRomConvergence(self):
+    """
+      This function will check the convergence of all ROMs
+      @ In, None
+      @ Out, bool, True if all ROMs are converged
+    """
+    converged = True
+    for romInfo in self.romsDictionary.values():
+      if not romInfo['Converged']:
+        converged = False
+    if converged:
+      self.raiseADebug("All ROMs are converged")
+    return converged
+
+  def checkRomValidity(self, kwargs):
+    """
+      This function will check the validity of all roms
+      @ In, kwargs, dict,  is a dictionary that contains the information coming from the sampler,
+        a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
+      @ Out, None
+    """
+    allValid = False
+    if self.modelSelection == 'CrowdingDistance':
+      allValid = self.crowdingDistanceMethod(kwargs)
+    else:
+      self.raiseAnError(IOError, "Unknown model selection method ", self.modelSelection, " is given!")
+
+    if allValid:
+      self.raiseADebug("ROMs  are all valid for given model ", self.modelInstance.name)
+
+    return allValid
+
+  def crowdingDistanceMethod(self, kwargs):
+    """
+      This function will check the validity of all roms based on the crowding distance method
+      @ In, kwargs, dict,  is a dictionary that contains the information coming from the sampler,
+        a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
+      @ Out, None
+    """
+    allValid = True
+    for romInfo in self.romsDictionary.values():
+      valid = False
+      # generate the data for input parameters
+      paramsList = romInfo['Instance'].getInitParams()['Features']
+      trainInput = np.asarray(self._extractInputs(romInfo['Instance'].trainingSet, paramsList).values())
+      currentInput = np.asarray(self._extractInputs(kwargs['SampledVars'], paramsList).values())
+      coeffCD = self.computeCDCoefficient(trainInput, currentInput)
+      self.raiseADebug("Crowding Distance Coefficient: ", coeffCD)
+      if coeffCD > 0.2:
```

Comment:
the coeffCD threshold should not be hardcoded (default is fine, but not hardcoded)

Reply:
changed.